### PR TITLE
[12.0][ADD] purchase_supplier_approve: supplier approval funcionality

### PR DESCRIPTION
**Pull request by dsidorenkovas-rgbconsulting**
This module adds an approval state for suppliers. There are 3 available states:

  * Pending to approval (by default)
  * Approved
  * Not approved

Suppliers can be approved only by a specific group of users: **Suppliers approval**.

The supplier approval state is used in purchase order confirmation process.
Only orders with approved suppliers can be confirmed.

### CODE REVIEW

**Comment by kittiu** [comment]
Thanks contribution :+1: 
To start with, can you change name of this PR to [12.0][ADD] purchase_supplier_approve
Also, please greenify the travis first. :)

File: purchase_supplier_approve/__manifest__.py
Comment (MeetKD) [inline purchase_supplier_approve/__manifest__.py]: Add OCA as author.

File: purchase_supplier_approve/models/res_partner.py
Comment (MeetKD) [inline purchase_supplier_approve/models/res_partner.py]: It would be great if you can add help string.

**Review by MeetKD — CHANGES_REQUESTED** [review]

File: purchase_supplier_approve/__manifest__.py
Comment (dsidorenkovas-rgbconsulting) [reply purchase_supplier_approve/__manifest__.py]: Hi,
As you can see I made this change in the following commit:
https://github.com/OCA/purchase-workflow/pull/744/commits/b6f823d3cfa9d4a71fd8cc4b6128cda08aeb6d33#diff-8e65f9724534767cee0c8449ebf1c8dcR9

File: purchase_supplier_approve/models/res_partner.py
Comment (dsidorenkovas-rgbconsulting) [reply purchase_supplier_approve/models/res_partner.py]: Hi again,
As you asked, I made this change also in the following commit:
https://github.com/OCA/purchase-workflow/pull/744/commits/b6f823d3cfa9d4a71fd8cc4b6128cda08aeb6d33#diff-d44cb370023ae01225035b3421763d4eR15

**Comment by dsidorenkovas-rgbconsulting** [comment]
> Thanks contribution 
> To start with, can you change name of this PR to [12.0][ADD] purchase_supplier_approve
> Also, please greenify the travis first. :)

Hi @kittiu,

I renamed the PR as you asked and also checked some travis errors. The current errors are caused by tests from other modules, since in this module we are blocking purchase order confirmation.

File: purchase_supplier_approve/models/res_partner.py
Comment (MeetKD) [reply purchase_supplier_approve/models/res_partner.py]: Looks good :+1:

File: purchase_supplier_approve/__manifest__.py
Comment (MeetKD) [reply purchase_supplier_approve/__manifest__.py]: :+1:

File: purchase_supplier_approve/__manifest__.py
Comment (kittiu) [inline purchase_supplier_approve/__manifest__.py:3]: This is a minor comment and non-blocking, but would you use single quote to all of your code? It is helpful when we do searhing for code later on.

File: purchase_supplier_approve/__manifest__.py
Comment (kittiu) [inline purchase_supplier_approve/__manifest__.py:18]: Do you plan to be maintainer of this module? (It will helps for you to easily merge it later).
'maintainers': ['your_git_login'],

File: purchase_supplier_approve/wizard/supplier_approval_wizard.py
Comment (kittiu) [inline purchase_supplier_approve/wizard/supplier_approval_wizard.py]: Track visibility is not needed I think, as this is a wizard

File: purchase_supplier_approve/wizard/supplier_approval_wizard.py
Comment (kittiu) [inline purchase_supplier_approve/wizard/supplier_approval_wizard.py]: Looks like, partner_id is not really needed, right? (you need it only because you want to set its default).
On open wizard, I would prefer to use `def default_get(...)` to set default supplier approval state (you can use _default_partner or _get_partner for partner object)
On save, again, you can still use _get_partner to get the partner and write it.

Note:
* This way, we can avoid using _onchange_partner_id, just to get the default approve state. To use method for it right purpose.
* Code can be shorten a bit

WDYT?

File: purchase_supplier_approve/models/res_partner.py
Comment (kittiu) [inline purchase_supplier_approve/models/res_partner.py]: I would prefer field name like, supplier_approve_status, or supplier_status, which easier to understand at first glance.
For selection, I would prefer,
[('pending_approve', 'Pending Approval'), ('approved', 'Approved'), ('not_approved', 'Not approved')]
My point is about readability. As such, when you see this column from database,
`supplier_approve_status = 'approved'`, would be easier to understand than, `supplier_approve = 'apd'`

File: purchase_supplier_approve/models/purchase.py
Comment (kittiu) [inline purchase_supplier_approve/models/purchase.py]: I seem many lines with length > 79 chars, it should be in new line.
I also wonder why it can pass travis test... my editor whould show error like this,
image

**Comment by dsidorenkovas-rgbconsulting** [comment]
Hi,

I made the requested changes. Let me know if there is anything else to be done.

**Comment by kittiu** [comment]
@dsidorenkovas-rgbconsulting this module does make test in other module failed with this,
```
You cannot confirm a purchase order if the supplier is not approved.
```
To coexists this module with others, is it ok to default state = 'approved' ? Will it still make sense?
Otherwise, I am not sure how this module can be merged...

**Comment by geluboros** [comment]
Hi @kittiu,

You are right, this module cannot coexist in this repository because it's making all the tests fail.

I think it's important to maintain "Pending Approval" as a default state, so maybe a solution could be to not block purchase confirmations for suppliers in this state.
We could add a system parameter flag that allows to configure if the pending state is a blocking state or not (False by default). Ofc, this should be well documented in the module's Readme to avoid any confusion.

What do you think about this approach? If you think it could work, let us know so we can apply the changes, otherwise this PR can be closed.

Regards,
Gelu

**Comment by kittiu** [comment]
Yes it is good opinion IMHO. Make it configurable :)

**Comment by dsidorenkovas-rgbconsulting** [comment]
Hi @kittiu,
I made the proposed changes. Also, I added another commit for child contacts managment.

**Pull request closed without merging**
